pydantic_ai_slim/pydantic_ai/_parts_manager.py

```diff
@@ -47,6 +47,75 @@
 """
 
 
+def _parse_chunk_for_thinking_tags(
+    content: str,
+    buffered: str,
+    start_tag: str,
+    end_tag: str,
+    in_thinking: bool,
+) -> tuple[list[tuple[str, str]], str]:
+    """Parse content for thinking tags, handling split tags across chunks.
+
+    Args:
+        content: New content chunk to parse
+        buffered: Previously buffered content (for split tags)
+        start_tag: Opening thinking tag (e.g., '<think>')
+        end_tag: Closing thinking tag (e.g., '</think>')
+        in_thinking: Whether currently inside a ThinkingPart
+
+    Returns:
+        (segments, new_buffer) where:
+        - segments: List of (type, content) tuples
+        - type: 'text'|'start_tag'|'thinking'|'end_tag'
+        - new_buffer: Content to buffer for next chunk (empty if nothing to buffer)
+    """
+    combined = buffered + content
+    segments: list[tuple[str, str]] = []
+    current_thinking_state = in_thinking
+    remaining = combined
+
+    while remaining:
+        if current_thinking_state:
+            if end_tag in remaining:
+                before_end, after_end = remaining.split(end_tag, 1)
+                if before_end:
+                    segments.append(('thinking', before_end))
+                segments.append(('end_tag', ''))
+                remaining = after_end
+                current_thinking_state = False
+            else:
+                # Check for partial end tag at end of remaining content
+                for i in range(len(remaining)):
+                    suffix = remaining[i:]
```

Comment:
We're creating a lot of suffixes that we'll immediate discard. I think we should be able to iterate only over the actual suffixes but using `remaining[-N:]` to look at the last N chars, and then fill for N use `1` through `len(end_tag)` (with a check for the length of remaining)